design/design.md

```diff
@@ -0,0 +1,251 @@
+function resolution.
+
+We will provide an example implementing a hash function with the right interface
+for ALBA using, for example, BLAKE underneath, but with a warning that
+it is provided for demonstration / prototyping only and should not be used in
+production.
+
+One reason is that when ALBA is used alongside other sub-protocols that also
+need hashing, it is the user's responsibility to implement
+[domain separation](https://en.wikipedia.org/wiki/Domain_separation)
+to provide the different
+sub-protocols with independent hash functions.
+Otherwise, the underlying assumptions needed for security might break and the
+larger protocol can become insecure
+(see [here](https://link.springer.com/chapter/10.1007/978-3-030-45724-2_1)).
+
+Additionally, different hash functions (e.g., SHA and BLAKE) shouldn't mix and
+match.
+The following example is contrived, but it demonstrates why using different
+hash function constructions within the same protocol can be insecure.
+Suppose $H$ and $H'$ are two distinct random looking hash functions.
+You might expect that the function $G$ given by $G(x) = H(x) \oplus H'(x)$,
+where $\oplus$ is the bitwise xor, is also random looking, but it's not true.
+Let $H$ be a hash function modeled as a random oracle and define
+$H(x) = \sim H'(x)$,
+where $\sim$ is bitwise not; then $H'$ is also random looking
+but $G(x) = 111...111$ for all $x$.
+Even if we use domain separation for $H$ and $H'$, we cannot guarantee that
+$G(x) = H(0, x) \oplus H'(1, x)$ is random looking.
+Define $H'(b, x) = H(\sim b, x)$; then $G(x) = 000...000$ for all $x$.
+
+## Generic data elements
+Similar to the above, we should also let data elements have a generic type.
+For example, the ALBA implementation should work with a signature type that
+exposes the underlying data as a `u8` slice.
+The exact interfaces will need to be determined.
+
+## Mathematical rigor and production-readiness
+At this point, it is not our goal to make the implementation 100% correct and
+secure.
+We need to roll out a prototype so that our primary clients (Peras, etc) can
+incorporate it in their prototypes, and also so that the external community can
+start playing with it as well.
+This means that there might be tiny rounding errors in parameter derivation due
+to the use of floats, and there might be tiny errors when using a uniformly
+random hash output to sample some distribution needed for ALBA
+(e.g., Bernoulli, Uniform, Binomial, Poisson).
+We will also probably not rigorously analyze the $H_1$ optimization described in
+section 3.3 of the paper.
+
+## Parameter calculation
+ALBA has two kinds of parameters: user friendly parameters such as
+$\lambda_{sec}$, $\lambda_{rel}$, $n_p$, $n_f$,
+and algorithm specific parameters such as $u$, $d$ and $q$.
+
+Our implementation will have parameter derivation code converting user friendly
+to algorithm specific parameters.
+However, this should be decoupled from the ALBA algorithms:
```

Comment:
I strongly disagree.
Users would typically want already made functions (with hardwritten security levels), think of SHA256, SHA512...
If we want people to "play" with our implementation, they should be able to put whatever they want. 
If we want them to "use" it, we should restrict them.
If you are worried about backward compability, users can simply fix the repository version.

Reply:
It is not ideal indeed, but leaving users the impressions that what they are doing is safe is usually worse.

I am talking about tags and releases on github (https://github.com/cardano-scaling/alba/tags) , as well as official versions when we publish on cargo